### PR TITLE
BIG-3884 - Port PSL

### PR DESCRIPTION
**Pull request by dlecocq**
Hacked this up over the weekend.

The interpretation of `url-py`, `moz-util-py` and this now all differ when asking for the PLD of a TLD:

```
# This returns the TLD, PLD, FQDN in reversed form
>>> matcher.get_all('co.uk')
('uk.co', 'uk.co', 'uk.co')
```

```
>>> URL.parse('http://co.uk/').pld
'co.uk'
>>> URL.parse('http://co.uk/').tld
'uk'
```

In this implementation, however, such a request raises an exception:

```
PSL psl = ...;

# This doesn't raise an exception
psl.getTLD("co.uk");

# This /does/ raise an exception
psl.getPLD("co.uk");
```

That was loosely based on the expectations set out in the [tests listed on the PSL site](https://raw.githubusercontent.com/publicsuffix/list/master/tests/test_psl.txt). Those tests would have us find `null` in exactly this case. I don't know what the most idiomatic thing to do in C++ for that case is, so until we figure that out, it throws an exception.

Oh, also, it works for both `punycoded` and `unpunycoded` forms without doing any transformations.

@b4hand @lindseyreno @tammybailey @neilmb

### CODE REVIEW

File: test/test-psl.cpp [hunk 1]
```diff
@@ -0,0 +1,751 @@
...
+{
+    std::string example = "WwW.example.COM";
+    std::string pld = "example.com";
+    std::string tld = "com";
+    EXPECT_EQ(pld, getPSL().getPLD(example));
+    EXPECT_EQ(tld, getPSL().getTLD(example));
+    EXPECT_EQ(std::make_pair(tld, pld), getPSL().getBoth(example));
+}
+
+/*
```
Comment (b4hand) [inline test/test-psl.cpp]: Why are these tests commented out?

Comment (dlecocq) [inline test/test-psl.cpp]: Because I've yet to make them pass.

**Comment by dlecocq** [comment]
Reupped

**Comment by b4hand** [comment]
LGTM

**Comment by dlecocq** [comment]
I guess my main question here is -- is raising an exception when there's no PLD the right thing to do?

**Comment by b4hand** [comment]
To be honest, I'm not sure. It's not very _exceptional_, and in C++ they don't lean on exceptions like in Python. However, I'm not sure what a better interface is. I don't like returning an empty string, and changing it to return by pointer is worse as well. If only we had the maybe monad. :)

**Comment by dlecocq** [comment]
Sounds like this is a question for @martin-seomoz , perhaps. Maybe we have a constant `Url::PSL::NotFound` that we return (like `std::string::npos`). One situation that I'd like to avoid is:

``` C++
// Neither really has a PLD, but they shouldn't have the _same_ PLD
psl.getPLD("com") == psl.getPLD("org")
```

**Comment by b4hand** [comment]
Unless you make it an entirely different type from `std::string`. There's no way to provide that guarantee. There's no string value such that is equivalent to `NaN`.

File: src/psl.cpp [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+namespace Url
+{
+
+    PSL::PSL(std::istream& stream)
+    {
+        std::string line;
+        while (std::getline(stream, line))
+        {
+            // Skip blank lines
+            if (!line.size())
```
Comment (ghost) [inline src/psl.cpp]: probably more descriptive to write `if (line.empty())`

File: src/psl.cpp [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            {
+                continue;
+            }
+
+            // Skip comments
+            if (line.size() >= 2 && line[0] == '/' && line[1] == '/')
```
Comment (ghost) [inline src/psl.cpp]: Easier to understand as `if (line.compare(0, 2, "//") == 0)`
But what about comment lines that don't begin as the first character?
`if (line.compare(line.find_first_not_of(" \t"), 2, "//") == 0)`

File: src/psl.cpp (hunk 2 above)
Comment (dlecocq) [inline src/psl.cpp]: Good point

File: src/psl.cpp (hunk 3 above)
Comment (dlecocq) [inline src/psl.cpp]: There should actually be another rule to set the line to be only up to the first whitespace.

But TIL about `int compare (size_t pos, size_t len, const string& str) const;`

**Pull request merged**